File: sdks/dotnet/ReproIt.Sdk/ManagedProjectToken.cs
namespace ReproIt.Sdk;

/// <summary>Restricts a project token to managed workload registration.</summary>
public sealed class ManagedProjectToken
{
    private const int MaxProjectTokenBytes = 1_024;
    private readonly string value;

    /// <summary>Validates and wraps one managed project token.</summary>
    public ManagedProjectToken(string value)
    {
        if (value.Length is 0 or > MaxProjectTokenBytes ||
            value.Any(character => character is < '!' or > '~'))
        {
            throw new ManagedCaptureException(
                "SCHEMA_INVALID", "The managed project token is invalid.");
        }
        this.value = value;
    }

    internal string SdkEngineValue => value;
}

File: sdks/go/reproit/managed_project_token.go
package reproit

const managedMaxProjectTokenBytes = 1_024

// ManagedProjectToken authorizes one managed workload registration.
type ManagedProjectToken struct {
	value string
}

// NewManagedProjectToken validates and wraps one managed project token.
func NewManagedProjectToken(value string) (*ManagedProjectToken, error) {
	if value == "" || len(value) > managedMaxProjectTokenBytes {
		return nil, newManagedError("SCHEMA_INVALID", "The managed project token is invalid.")
	}
	for index := 0; index < len(value); index++ {
		if value[index] < 33 || value[index] > 126 {
			return nil, newManagedError("SCHEMA_INVALID", "The managed project token is invalid.")
		}
	}
	return &ManagedProjectToken{value: value}, nil
}

func (token *ManagedProjectToken) sdkEngineValue() string {
	if token == nil {
		return ""
	}
	return token.value
}
